src/operator/subgraph/static_shape_subgraph_property.cc

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "./common.h"
+#include "./subgraph_property.h"
+#include "../../imperative/cached_op.h"
+
+namespace mxnet {
+namespace op {
+
+/*
+ * This selects nodes for a subgraph that only contains static shape operators
+ * and it visits nodes via both input and output links.
+ */
+class StaticShapeOpSelector: public SubgraphSelector {
+ public:
+  // select nodes with FInferShape attribute
+  bool Select(const nnvm::Node &seed_node) override {
+    const auto& infershape = nnvm::Op::GetAttr<mxnet::FInferShape>("FInferShape");
+    return !seed_node.is_variable() && infershape.count(seed_node.op()) &&
+           !unsupported_op_names_.count(seed_node.op()->name);
+  }
+
+  bool SelectInput(const nnvm::Node &cur_node, const nnvm::Node &input_node) override {
+    return Select(input_node);
+  }
+
+  bool SelectOutput(const nnvm::Node &cur_node, const nnvm::Node &output_node) override {
+    return Select(output_node);
+  }
+
+  // reject single node subgraph
+  std::vector<nnvm::Node*> Filter(const std::vector<nnvm::Node*>& candidates) override {
+    if (candidates.size() == 1) {
+      return std::vector<nnvm::Node*>();
+    }
+    return candidates;
+  }
+
+ private:
+    // static shape ops that fail backward pass inside subgraph CachedOp
+    // GitHub issue: https://github.com/apache/incubator-mxnet/issues/18823
+    std::unordered_set<std::string> unsupported_op_names_ {"Reshape", "_np_reshape", "transpose",
```

Comment:
what did you guys decide to do about these fixed ops to exclude? Is it worth having an option for users to specify a list of ops to exclude, or is that something we should consider in the future based on the feedback you get after this PR?

Since you added the top level partition_if_dynamic option, users can always turn it off if it doesnt work correctly. So that should cover the case where this change breaks existing use cases (if at all).

Reply:
You put it correctly, we can decide on providing that option based on feedback from users.